### ISSUE TITLE
file.managed doesn't work with ldap users

### ISSUE THREAD

**Issue by samueletonon** [saltstack/salt#12803]
scenario:
user samu comes from an ldap:

```
$id samu
uid=10000(samu) gid=10000(utenti)
```

let's have this testing.sls file:

```
/home/samu/testing:
  file.managed:
    - source: salt://testing
    - mode: 640
    - user: samu
    - group: utenti
```

and let's run it, it will give you the following error:

```
$salt '*' state.highstate testing.sls
----------
    State: - file
    Name:      /home/samu/testing
    Function:  managed
        Result:    False
        Comment:   User samu is not available Group utenti is not available
        Changes:
```

**Comment by jessegonzalez** [comment]
I've noticed that calling 'salt-call state.{sls,highstate,...} [state_file1[,...]]' on the minion directly works as expected.

**Comment by lobosky** [comment]
At which point is now the LDAP support for file.* ? 
Thanks.

**Comment by babilen** [comment]
Proper LDAP (or rather: full PAM) support would really be great as it is being used in many setups that now require awkward workarounds.

**Comment by andrewmeyer** [comment]
+1, this is blocking deployments at my company

**Comment by basepi** [comment]
I've labeled this for potential inclusion in the Boron release of salt. No promises, just wanted to get eyes on it.

**Comment by rhansen** [comment]
I've only experienced problems using LDAP users/groups with `file.managed` in one particular case:  If I bring up a brand new machine and run `state.highstate` on it, then the minion fails to see LDAP users/groups until enough time has elapsed or I restart the minion process.

I believe this happens because glibc doesn't check `/etc/nsswitch.conf` for changes very often (ever?) once it has read it.  I've noticed that an existing long-running process (like `salt-minion`) will start seeing the LDAP users after a few minutes, but I'm not sure why.  Maybe glibc periodically re-reads `/etc/nsswitch.conf`, or maybe it's because my `state.highstate` installs `nscd` and glibc periodically tries connecting to `nscd`.

To work around this problem I created a custom state module that has a function that calls `__salt__['file.user_to_uid']()` and `__salt__['file.group_to_gid']()` every second until a known user and group appears.  I then modified my `file.managed` entries to depend on that function.

See this old (and possibly no longer accurate) bug report:  https://bugzilla.redhat.com/show_bug.cgi?id=132608

**Comment by grichmond-salt** [comment]
Team,

Adding the following notes from a PS discussion:

```
Observation
-----------
User is experiencing an LDAP user and group lookup issue on the Minion.

Investigation
-------------
The following GitHub Issue describes the customer’s observations:
https://github.com/saltstack/salt/issues/12803 (this one)

Modules Involved
----------------
Non Local Users (LDAP) not working in the following modules:
- file.managed
- file.directory
- cmd.script

Workarounds
-----------
a) File Module:
- specified UID and GID for file modules
b) Command Module:
- omitted "user" and "group" options
- used 'su - ldap_username -c "path_to_script"' as an alternative

Long Term Request
-----------------
User would prefer to leverage user and group in the usual manner, even though the UID and GID are provided by LDAP.  This would be true for any built-in module where the options of " - user:" or "- group:" are present.
```

**Comment by mallniya** [comment]
+1, also blocking deployment

**Comment by moloney** [comment]
I am running into this issue as well (on Ubuntu 16.04).  I had to restart the salt minion process for it to start recognizing users/groups from LDAP.  No amount of waiting makes any difference. 

I am pretty sure that the redhat bug linked to by @rhansen describes the issue we are hitting. I took a look at the salt code and noticed it is using the 'pwd' and 'grp' modules from the standard library, which in turn call the glibc functions 'getgr*' and 'getpw*'. I actually do install nscd after setting up nsswitch, but since the salt minion process was started before that it doesn't help.

As a quick note, this has nothing to do with PAM.  It is common to setup nsswitch to get user/group info from LDAP without setting up PAM to allow those users to login (e.g. on an NFS file server).

**Comment by knine** [comment]
+1, also running into this issue

**Comment by cavepopo** [comment]
+1, also running into this issue

**Comment by wimvds** [comment]
+1, just ran into this issue. Restarting the salt-minion service as mentioned by @moloney (and running state.highstate again) did indeed solve the issue for me (minion is a Debian 8 system).

**Comment by johnkeates** [comment]
Ran into the same issue here. Debian 8 too. Restarting it resolved it for a bit here as well.

**Comment by kalmeshcs** [comment]
+1, also running into this issue
It works fine when I run the individual salt state from the minion (using salt-call --local state.sls) but doesn't seem to work when this state is triggered from master.

**Comment by L4rS6** [comment]
+1, just ran into this issue

**Comment by justin-sunayu** [comment]
+1 Triggering scripts to work around this.

**Comment by vrubiella** [comment]
+1 same here

**Comment by mysticmoron** [comment]
+1 restart does not work for me either on Centos 7. 

The minion is able to join the ldap fine during the highstate itself, other scripts work fine with remote user/group.
And then I am manually applying a state from salt-master, but `file.directory` simply can not locate either the ldap user or group.

**Comment by wegenerbenjamin** [comment]
This issue may be solved with glibc >= 2.33, see https://developers.redhat.com/blog/2021/01/15/coming-in-glibc-2-33-reloadable-nsswitch-conf (not yet tested)

**Comment by thatch45** [comment]
Given that this should have been fixed with an upstream glibc issue and has not been verified on a supported release I am going to close it out.
If this can be re-verified on a current release please open a fresh issue that follows current issue filing standards.
